pkg/runner/mounting_runner.go

```diff
@@ -0,0 +1,105 @@
+package runner
+
+import (
+	"context"
+	"os"
+	"strings"
+
+	"github.com/buildbarn/bb-remote-execution/pkg/proto/configuration/bb_runner"
+	runner_pb "github.com/buildbarn/bb-remote-execution/pkg/proto/runner"
+	"github.com/buildbarn/bb-storage/pkg/filesystem"
+	"github.com/buildbarn/bb-storage/pkg/filesystem/path"
+	"github.com/buildbarn/bb-storage/pkg/util"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/emptypb"
+)
+
+type mountingRunner struct {
+	base        runner_pb.RunnerServer
+	buildDirectory filesystem.Directory
+	mount *bb_runner.MountInfo
+}
+
+// NewMountingRunner is a decorator for Runner
+// that mounts `mount` before running a build action.
+//
+// This decorator can be used for chroot runners
+// that must mount special filesystems into the input root.
+func NewMountingRunner(base runner_pb.RunnerServer, buildDirectory filesystem.Directory, mount *bb_runner.MountInfo) runner_pb.RunnerServer {
+	return &mountingRunner{
+		buildDirectory: buildDirectory,
+		mount: mount,
+		base: base,
+	}
+}
+
+func open(buildDirectory filesystem.Directory, inputRootDirectory string) (filesystem.DirectoryCloser, error) {
+	// TODO(nils): How to best open the input root `Directory`
+	//             It must be a `localDirectory`, but outer directories can be `lazyDirectory`.
+	// NB: We iterate over Directory, as the initial buildDirectory is not a DirectoryCloser.
+	iterator := buildDirectory
+	var dir filesystem.DirectoryCloser
+
+	parts := strings.Split(inputRootDirectory, "/")
+	if len(parts) == 0 {
```

Comment:
The input root directory is in theory permitted to be empty. There is nothing in the runner protocol that forbids it. It may also be ".".